src/client/Microsoft.Identity.Client/ManagedIdentity/KeyProviders/WindowsManagedIdentityKeyProvider.cs

```diff
@@ -0,0 +1,154 @@
+using System;
+using System.Security.Cryptography;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client.ManagedIdentity.KeyGuard;
+
+namespace Microsoft.Identity.Client.ManagedIdentity.Providers
+{
+    /// <summary>
+    /// Windows policy:
+    ///   1) KeyGuard (CVM/TVM) if available
+    ///   2) Hardware (TPM/KSP via Microsoft Platform Crypto Provider)
+    ///   3) In-memory fallback (delegates to InMemoryManagedIdentityKeyProvider)
+    /// No certs; no attestation; no long-lived handles kept.
+    /// </summary>
+    internal sealed class WindowsManagedIdentityKeyProvider : IManagedIdentityKeyProvider
+    {
+        private static readonly SemaphoreSlim s_once = new SemaphoreSlim(1, 1);
+        private volatile MiKeyInfo _cached;
+
+        private const string KgProviderName = "Microsoft Software Key Storage Provider";
+        private const string KgKeyName = "KeyGuardRSAKey";
+        private const string TpmProvider = "Microsoft Platform Crypto Provider";
+        private const string TpmKeyName = "MSAL_MI_PLATFORM_RSA";
+
+        public async Task<MiKeyInfo> GetOrCreateKeyAsync(CancellationToken ct)
+        {
+            if (_cached != null)
+                return _cached;
+
+            await s_once.WaitAsync(ct).ConfigureAwait(false);
+            try
+            {
+                if (_cached != null)
+                    return _cached;
+
+                // 1) KeyGuard (RSA-2048 under VBS isolation)
+                RSA kgRsa;
+                if (TryGetOrCreateKeyGuard(out kgRsa))
+                {
+                    _cached = new MiKeyInfo(kgRsa, MiKeyType.KeyGuard);
+                    return _cached;
+                }
+
+                // 2) Hardware TPM/KSP (RSA-2048, non-exportable)
+                RSA hwRsa;
+                if (TryGetOrCreateHardwareRsa(out hwRsa))
+                {
+                    _cached = new MiKeyInfo(hwRsa, MiKeyType.Hardware);
+                    return _cached;
+                }
+
+                // 3) Delegate fallback to portable in-memory provider
+                var memProvider = new InMemoryManagedIdentityKeyProvider();
+                _cached = await memProvider.GetOrCreateKeyAsync(ct).ConfigureAwait(false);
+                return _cached;
+            }
+            finally
+            {
+                s_once.Release();
+            }
+        }
+
+        // --- KeyGuard path (RSA) ---
+        private static bool TryGetOrCreateKeyGuard(out RSA rsa)
+        {
+            rsa = default(RSA);
+
+            try
+            {
+                CngProvider provider = new CngProvider(KgProviderName);
+
+                CngKey key;
+                if (CngKey.Exists(KgKeyName, provider))
+                {
+                    key = CngKey.Open(KgKeyName, provider);
+
+                    // Ensure actually KeyGuard-protected; if not, recreate as KeyGuard.
+                    if (!KeyGuardKey.IsKeyGuardProtected(key))
```

Comment:
Is it worth adding all the `KeyGuardKey` methods in this class?

Reply:
add package reference now and removed the conditions